import { Link } from 'src/components/Link/Link'
import { LinkExternal } from 'src/components/Link/LinkExternal'
import VoC from 'content/clusters/VoCHeader.md'

<VoC/>

### 21K (Omicron)
Also known as <Lin name="BA.1" /> and <Who name="Omicron" />

<i>
<Var name="21K (Omicron)" prefix=""/> is part of a larger group, <b>21M (Omicron)</b>, which corresponds to Pango lineage <Lin name="B.1.1.529" />. <b>21M (Omicron)</b> includes <Var name="21K (Omicron)" prefix=""/>, its sister-clade <Var name="21L (Omicron)" prefix=""/>, and other diverse Omicron sequences. See the <Var name="21L (Omicron)" prefix=""/> page for more information.
</i>
<br/><br/>

<Var name="21K (Omicron)"/> appears to have arisen in November 2021, possibly in South Africa. Early sequences are predominantly from South Africa, though also detected in Botswana and Hong Kong.
<br/><br/>

<Var name="21K (Omicron)" prefix=""/> is primarily of concern due to the large number of mutations it has in the Spike gene. Many of these variants are in the receptor binding domain and N-terminal domain, and thus may play key roles in ACE2 binding and antibody recognition.
<br/><br/>

The deletion at positions <AaMut mut={'S:H69-'}/> and <AaMut mut={'S:V70-'}/> (see <Mut name="S:H69-"/>)) is also found in <Var name="20I (Alpha, V1)" prefix=""/> and <Var name="21D (Eta)" prefix=""/>, along with some other variants. The 69/70 deletion in this variant causes the S-assay within TaqPath tests to give a negative result, which can provide a useful proxy for prevalence of this variant (a phenomenon referred to as S-gene target failure or SGTF). However, as this deletion is found in other variants, care must be taken to ensure other variants aren't being detected.

There is a 3 amino-acid insertion in Spike of <code>'EPE'</code> at position 214.

A particular cluster of mutations at the S1-S2 furin cleavage site (<AaMut mut={'S:H655Y'}/>, <AaMut mut={'S:N679K'}/>, <AaMut mut={'S:P681H'}/>) may also be associated with increased transmissibility [<LinkExternal href="https://www.biorxiv.org/content/10.1101/2021.08.04.455140v1">Yu et al., bioRxiv</LinkExternal>).
The mutation at <AaMut mut={'S:E484A'}/> is at a position that's also mutated in other variants (see <Mut name="S:E484"/>)), and mutations to other amino-acids at this position have been associated with immune-escape ([Wise, BMJ](https://www.bmj.com/content/372/bmj.n359)).

The combination of mutations <AaMut mut={'S:Q498R'}/> and <AaMut mut={'S:N501Y'}/> in in-vitro evolution studies signifciantly increased the binding affinity to ACE2 ([Zahradnik et al., Nature Micribiology](https://www.nature.com/articles/s41564-021-00954-4)).

Note that the 3 amino-acid deletion and mutation at <AaMut mut={"S:G142-"}/> <AaMut mut={"S:V143-"}/> <AaMut mut={"S:Y144-"}/> <AaMut mut={"S:Y145D"}/> is sometimes reported by others as <AaMut mut={"S:G142D"}/> <AaMut mut={"S:V143-"}/> <AaMut mut={"S:Y144-"}/> <AaMut mut={"S:Y145-"}/>. This is simply a matter of where one places the deletions and the mutation. CoVariants follows the alignment patterns of Nextstrain/Nextclade so that what's reported here is reflected in what is seen in the focal builds. Similarly, the deletion and mutation at <AaMut mut={"S:N211-"}/> <AaMut mut={"S:N212I"}/> is sometimes reported as <AaMut mut={"S:N211I"}/> <AaMut mut={"S:N212-"}/>. Again, this simply reflects a different positioning of the deletions. In both cases, the resulting protein sequences are identical. 

Additionally, there is a 3 amino-acid deletion in <code>ORF1a</code> at <AaMut mut={'ORF1a:L3674-'}/>, <AaMut mut={'ORF1a:S3675-'}/>, and <AaMut mut={'ORF1a:G3676-'}/> (this is also annotated as a deletion in NSP6 from 105-107). There is some speculation that this mutation could aid in innate immune evasion, possibly by compromising cells' ability to degrade viral components ([Benvenuto et al., Journal of Infection](https://www.sciencedirect.com/science/article/pii/S0163445320301869)).

The accessory protein <code>ORF9b</code> has a 3 amino-acid deletion at <AaMut mut={'ORF9b:E27-'}/>, <AaMut mut={'ORF9b:N28-'}/>, and <AaMut mut={'ORF9b:A29-'}/>. It is believed that <code>ORF9b</code> is involved in suppression of the innate immune response to viral infection by interaction with TOM70 and NEMO and ultimately IFN production. ([Bayer et al., J. Mol. Biol.](https://doi.org/10.1016/j.jmb.2021.167265), [Thorn et al., bioRxiv](https://doi.org/10.1101/2021.06.06.446826)). It has been shown experimentally that <code>ORF9b</code> interacts with NEMO via it's N terminus region and that removal of the first 30 residues inhibits interaction ([Wu et al., Cell Reports](https://doi.org/10.1016/j.celrep.2021.108761)).


There are also two mutations in Nucleocapsid, <AaMut mut={'N:R203K'}/> and <AaMut mut={'N:G204R'}/>. Though these are ancestral (not new to this variant), they have been shown to be linked to increase subgenomic RNA expression ([Leary et al., bioRxiv](https://www.biorxiv.org/content/10.1101/2020.04.10.029454v2)) and increased viral loads ([Mourier et al., medRxiv](https://www.medrxiv.org/content/10.1101/2021.05.06.21256706v2)).

Please note that non-synonymous mutations (in the right-hand menu) have not yet been calculated - this will be updated as soon as sequences are finished processing.

- Prior to the variant's official designation as <Who name="Omicron" />, some reports prematurely referred to the variant as 'Nu,' which would have have been the next letter after <Who name="Mu" />. WHO apparently skipped 'Nu' because of confusion with the English word 'new' and 'Xi' because it is a common surname ([Kai Kupferschmidt, Twitter](https://twitter.com/kakape/status/1464671345070186497))
- WHO Technical Advisory Group on SARS-CoV-2 designed Omicron a variant of concern on 26 Nov 21 ([article](https://www.who.int/news/item/26-11-2021-classification-of-omicron-(BA.1)-sars-cov-2-variant-of-concern))
- Isolated Omicron viruses showed incomplete immune escape (41x reduction in titers) from antibodies induced by two doses of the Pfizer vaccine in individuals without prior SARS-CoV-2 infection, but antibody titers induced by two doses of the Pfizer vaccine in individuals with prior SARS-CoV-2 infection remained above levels expected to be protective ([Africa Health Research Institute](https://www.ahri.org/omicron-incompletely-escapes-immunity-induced-by-the-pfizer-vaccine/))
- Neutralization was observed to be completely lost for isolated Omicron viruses in sera from individuals vaccainated with 2 dose of Pfizer/BioNTech or Moderna, or heterologous AstraZeneca/BioNTech full vaccination. However, this was recovered to some degree with a booster shot or infection (<LinkExternal href="https://www.medrxiv.org/content/10.1101/2021.12.07.21267432v2">Wilhelm et al. medRxiv</LinkExternal>)
- In a pseudovirus assay with synthetic Omicron-like spike proteins, titer reduction was more variable - between 0x and 25x reduction in neutralization. (<LinkExternal href="https://drive.google.com/file/d/1CuxmNYj5cpIuxWXhjjVmuDqntxXwlfXQ/view">Sheward et al., Google Drive</LinkExternal>)


_Please help by providing links to further information about this variant if you can!_
