import { Link } from 'src/components/Link/Link'
import { LinkExternal } from 'src/components/Link/LinkExternal'

Also known as <Lin name="B.1.1.529" />

<Var name="21K"/> appears to have arisen in November 2021, possibly in South Africa. Early sequences are predominantly from South Africa, though also detected in Botswana and Hong Kong.
<br/><br/>

<Var name="21K" prefix=""/> is primarily of concern due to the large number of mutations it has in the Spike gene. Many of these variants are in the receptor binding domain and N-terminal domain, and thus may play key roles in ACE2 binding and antibody recognition.
<br/><br/>

The deletion at positions <AaMut mut={'S:H69-'}/> and <AaMut mut={'S:V70-'}/> (see <Mut name="S:H69-"/>)) is also found in <Var name="20I (Alpha, V1)" prefix=""/> and <Var name="21D (Eta)" prefix=""/>, along with some other variants. The 69/70 deletion in this variant causes the S-assay within TaqPath tests to give a negative result, which can provide a useful proxy for prevalence of this variant (a phenomenon referred to as S-gene target failure or SGTF). However, as this deletion is found in other variants, care must be taken to ensure other variants aren't being detected.

There is a 3 amino-acid insertion in Spike of <code>'EPE'</code> at position 214.

A particular cluster of mutations at the S1-S2 furin cleavage site (<AaMut mut={'S:H655Y'}/>, <AaMut mut={'S:N679K'}/>, <AaMut mut={'S:P681H'}/>) may also be associated with increased transmissibility ([Yu et al., bioRxiv](https://www.biorxiv.org/content/10.1101/2021.08.04.455140v1)).

The combination of mutations <AaMut mut={'S:Q498R'}/> and <AaMut mut={'S:N501Y'}/> in in-vitro evolution studies signifciantly increased the binding affinity to ACE2 ([Zahradnik et al., Nature Micribiology](https://www.nature.com/articles/s41564-021-00954-4)).

Additionally, there is a 3 amino-acid deletion in <code>ORF1a</code> at <AaMut mut={'ORF1a:L3674-'}/>, <AaMut mut={'ORF1a:S3675-'}/>, and <AaMut mut={'ORF1a:G3676-'}/> (this is also annotated as a deletion in NSP6 from 105-107). There is some speculation that this mutation could aid in innate immune evasion, possibly by compromising cells' ability to degrade viral components ([Benvenuto et al., Journal of Infection](https://www.sciencedirect.com/science/article/pii/S0163445320301869)).

There are also two mutations in Nucleocapsid, <AaMut mut={'N:R203K'}/> and <AaMut mut={'N:G204R'}/>. Though these are ancestral (not new to this variant), they have been shown to be linked to increase subgenomic RNA expression ([Leary et al., bioRxiv](https://www.biorxiv.org/content/10.1101/2020.04.10.029454v2)) and increased viral loads ([Mourier et al., medRxiv](https://www.medrxiv.org/content/10.1101/2021.05.06.21256706v2)).

Please note that non-synonymous mutations (in the right-hand menu) have not yet been calculated - this will be updated as soon as sequences are finished processing.


_Please help by providing links to further information about this variant if you can!_

